text/0432-contextual-helpers.md

```diff
@@ -0,0 +1,811 @@
+
+  <div class={{on-click}} />
+
+  <div class={{on-click "extra" "args"}} />
+
+  <div class={{on-click mouseenter=(action "highlight") mouseleave=(action "unhighlight")}} />
+
+  {{!-- runtime error: not a component --}}
+  <on-click />
+{{/let}}
+```
+
+### Relationship with globals
+
+Today, Ember apps rely heavily on the global namespace as the main mechanism of
+making components, helpers and modifiers available. Ideally, in a world where
+"everything is a value", the global and local namespace should behave the same
+way. Global components, helpers and modifiers should just be global variables
+that are implicitly defined around every templates in the app.
+
+In other words, it is as if every template has the following hidden "prelude"
+around its content:
+
+```hbs
+{{!-- prelude --}}
+{{#let (component "input") as |input|}}
+  {{#let (helper "concat") as |concat|}}
+    {{#let (modifier "action") as |action|}}
+      {{!-- ...other global components, helpers and modifiers omitted... --}}
+
+      {{!-- begin template content --}}
+      Your name:
+      {{concat this.firstName " " this.lastName}}
+
+      Change it:
+      {{input value=this.firstName}}
+      {{input value=this.lastName}}
+
+      <button {{action "submit"}}></button>
+      {{!-- end tempalte content ---}}
+    {{/let}}
+  {{/let}}
+{{/let}}
+```
+
+While this largely matches how things work today, there are a few notable
+differences where globals behave "unexpectedly" in this world.
+
+First of all, it is not possible to reference a component, helper or modifier
+in templates without invoking them today:
+
+```hbs
+{{!-- if `join-words` is a global helper, this works as expected --}}
+{{!-- this invokes the helper and yield the result --}}
+
+{{yield (join-words "foo" "bar" separator=",")}}
+         ~~~~~~~~~~
+
+{{!-- however, in this position, Ember does not "see" the helper --}}
+{{!-- this falls back to looking up the `join-words` property on `this` --}}
+
+{{yield join-words}}
+        ~~~~~~~~~~
+
+{{!-- as opposed to a "true" variable/value binding... --}}
+{{!-- this yields the helper as a value, as expected --}}
+
+{{#let (helper "join-words") as |join-words|}}
+  {{yield join-words}}
+          ~~~~~~~~~~
+{{/let}}
+```
+
+In the long term, we poropose to unify the semantics such that globals will
+behave exactly like local bindings (i.e. we should make this second case work).
+
+However, is not possible in the short term. This is due to the ambiguity
+between referencing a global variable and the
+[property lookup fallback](https://github.com/emberjs/rfcs/blob/master/text/0308-deprecate-property-lookup-fallback.md)
+feature. We propose we simply wait until the property lookup fallback is fully
+deprecated and removed, at which point we can reclaim the syntax.
+
+In the mean time, globals can be referenced explicitly using the `component`,
+`helper`, and `modifier` helpers.
+
+Another difference is how global helpers can be invoked without arguments in
```

Comment:
I think it's important to call out that this isn't about named arguments generally it's about named arguments to angle bracket invocations. 

Specifically, this does not invoke `pi` (it passes the value):

```hbs
{{my-component value=pi}}
```